File: python/847_访问所有节点的最短路径_hard.py
class Solution:
    def shortestPathLength(self, graph: List[List[int]]) -> int:
        n, ans = len(graph), 0
        q = deque((i, 1<<i, 0) for i in range(n))
        seen = {(i, 1<<i) for i in range(n)}
        
        while q:
            u, mask, dist = q.popleft()
            if mask == (1<<n) - 1:
                ans = dist
                break
            for v in graph[u]:
                mask_v = mask | 1<<v
                if (v, mask_v) not in seen:
                    q.append((v, mask_v, dist+1))
                    seen.add((v, mask_v))
        return ans

File: cpp/847_访问所有节点的最短路径_hard.cpp
class Solution {
public:
    int shortestPathLength(vector<vector<int>>& graph) {
        int n = graph.size();
        queue<tuple<int, int, int>> q;
        vector<vector<bool>> seen(n, vector<bool>(1<<n));
        int ans = 0;
        for (int i=0; i<n; i++) {
            q.emplace(i, 1<<i, 0);
            seen[i][1<<i] = true;
        }
        while (!q.empty()) {
            auto [u, mask, dist] = q.front();
            q.pop();
            if (mask == (1<<n) - 1) {
                ans = dist;
                break;
            }
            for (auto v : graph[u]) {
                int mask_v = mask | 1<<v;
                if (!seen[v][mask_v]) {
                    q.emplace(v, mask_v, dist+1);
                    seen[v][mask_v] = true;
                }
            }
        }
        return ans;
    }
};